icd_windows_dxgk.c

```diff
@@ -0,0 +1,145 @@
+        {
+            FreeLibrary(h);
+            return FALSE;
+        }
+        pAdapterInfo = (D3DKMT_ADAPTERINFO*)malloc(sizeof(D3DKMT_ADAPTERINFO)*(EnumAdapters.NumAdapters));
+        EnumAdapters.pAdapters = pAdapterInfo;
+        Status = pEnumAdapters2(&EnumAdapters);
+        if (!NT_SUCCESS(Status))
+        {
+            FreeLibrary(h);
+            if (pAdapterInfo) free(pAdapterInfo);
+            return FALSE;
+        }
+        for (UINT AdapterIndex = 0; AdapterIndex < EnumAdapters.NumAdapters; AdapterIndex++)
+        {
+            D3DDDI_QUERYREGISTRY_INFO QueryArgs = {0};
+            D3DDDI_QUERYREGISTRY_INFO* pQueryArgs = &QueryArgs;
+            D3DDDI_QUERYREGISTRY_INFO* pQueryBuffer = NULL;
+            QueryArgs.QueryType = D3DDDI_QUERYREGISTRY_ADAPTERKEY;
+            QueryArgs.QueryFlags.TranslatePath = TRUE;
+            QueryArgs.ValueType = REG_MULTI_SZ;
+#ifdef _WIN64
+            wcscpy_s(QueryArgs.ValueName, ARRAYSIZE(L"OpenCLDriverName"), L"OpenCLDriverName");
+#else
+            wcscpy_s(QueryArgs.ValueName, ARRAYSIZE(L"OpenCLDriverNameWow"), L"OpenCLDriverNameWow");
+#endif
+            D3DKMT_QUERYADAPTERINFO QueryAdapterInfo = {0};
+            QueryAdapterInfo.hAdapter = pAdapterInfo[AdapterIndex].hAdapter;
+            QueryAdapterInfo.Type = KMTQAITYPE_QUERYREGISTRY;
+            QueryAdapterInfo.pPrivateDriverData = &QueryArgs;
+            QueryAdapterInfo.PrivateDriverDataSize = sizeof(QueryArgs);
+            Status = D3DKMTQueryAdapterInfo(&QueryAdapterInfo);
+            if (!NT_SUCCESS(Status))
+            {
+                FreeLibrary(h);
+                if (pAdapterInfo) free(pAdapterInfo);
+                return FALSE;
+            }
+            if (NT_SUCCESS(Status) && pQueryArgs->Status == D3DDDI_QUERYREGISTRY_STATUS_BUFFER_OVERFLOW)
+            {
+                unsigned int QueryBufferSize = sizeof(D3DDDI_QUERYREGISTRY_INFO) + QueryArgs.OutputValueSize;
+                pQueryBuffer = (D3DDDI_QUERYREGISTRY_INFO*)malloc(QueryBufferSize);
+                memcpy(pQueryBuffer, &QueryArgs, sizeof(D3DDDI_QUERYREGISTRY_INFO));
+                QueryAdapterInfo.pPrivateDriverData = pQueryBuffer;
+                QueryAdapterInfo.PrivateDriverDataSize = QueryBufferSize;
+                Status = D3DKMTQueryAdapterInfo(&QueryAdapterInfo);
+                pQueryArgs = pQueryBuffer;
+            }
+            if (NT_SUCCESS(Status) && pQueryArgs->Status == D3DDDI_QUERYREGISTRY_STATUS_SUCCESS)
+            {
+                wchar_t* pWchar = pQueryArgs->OutputString;
+                unsigned int i = 0;
+                memset(cszLibraryName, 0, sizeof(cszLibraryName));
+                {
+                    while ((*pWchar != L'\0') && (i<1023))
+                    {
+                        cszLibraryName[i] = (char)*pWchar;
+                        i++;
+                        pWchar++;
+                    }
+                    if (i < 1023) khrIcdVendorAdd(cszLibraryName);
```

Comment:
Same goes here for 1023, and we should put the khrIcdVendorAdd on the following line.